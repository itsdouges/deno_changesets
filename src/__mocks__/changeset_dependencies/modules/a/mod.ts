import 'https://deno.land/x/b/mod.ts';
import 'https://deno.land/x/b@0.1.0/mod.ts';
import 'https://deno.land/x/b@0.2.0/main.ts';
import 'https://deno.land/x/b@v0.2.0/main.ts';
import 'https://deno.land/x/b@0.1.2/utils/mod.ts';
