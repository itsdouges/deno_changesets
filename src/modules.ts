import { join } from 'https://deno.land/std@0.170.0/path/mod.ts';
import { name } from './git.ts';

const topLevelModuleNames = /(main|index|mod)\.(js|ts)x?$/;

export interface Module {
  name: string;
  path: string;
}

export interface Repository {
  type: 'multi' | 'single';
  modules: Module[];
}

export async function list(
  path: string,
): Promise<Repository> {
  const foundModules: string[] = [];
  const stats = {
    hasTopLevelModule: false,
    hasModulesFolder: false,
  };

  for await (const dirEntry of Deno.readDir(path)) {
    if (dirEntry.isFile && topLevelModuleNames.exec(dirEntry.name)) {
      stats.hasTopLevelModule = true;
    }

    if (dirEntry.isDirectory && dirEntry.name === 'modules') {
      stats.hasModulesFolder = true;

      for await (const dirEntry of Deno.readDir(join(path, 'modules'))) {
        if (!dirEntry.isDirectory) {
          throw new Error(
            'invariant: children of modules folder must be folders',
          );
        }

        if (dirEntry.isFile && topLevelModuleNames.exec(dirEntry.name)) {
          stats.hasTopLevelModule = true;
        }

        foundModules.push(dirEntry.name);
      }
    }
  }

  if (stats.hasModulesFolder && stats.hasTopLevelModule) {
    throw new Error(
      'repository should have either a top level module or nested ones inside a modules folder, but not both.',
    );
  }

  if (stats.hasTopLevelModule) {
    return {
      type: 'single',
      modules: [{ name: await name(), path: path.replace(Deno.cwd(), '') }],
    };
  }

  if (stats.hasModulesFolder) {
    return {
      type: 'multi',
      modules: foundModules.map((mod) => ({
        name: mod,
        path: join(path.replace(Deno.cwd(), ''), 'modules', mod),
      })),
    };
  }

  throw new Error('invariant');
}
