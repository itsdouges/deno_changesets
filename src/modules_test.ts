import {
  assertEquals,
  assertRejects,
} from 'https://deno.land/std@0.170.0/testing/asserts.ts';
import { assertSnapshot } from 'https://deno.land/std@0.170.0/testing/snapshot.ts';
import { join } from 'https://deno.land/std@0.170.0/path/mod.ts';
import { list, updateVersion } from './modules.ts';

Deno.test(async function shouldReturnSingleModule() {
  const dir = join(Deno.cwd(), '/src/__mocks__/single');

  const actual = await list(dir);

  assertEquals(actual, {
    type: 'single',
    modules: [{ name: 'deno_changesets', path: '/src/__mocks__/single' }],
  });
});

Deno.test(async function shouldReturnMultiModule() {
  const dir = join(Deno.cwd(), '/src/__mocks__/multi');

  const actual = await list(dir);

  assertEquals(actual, {
    type: 'multi',
    modules: [{
      name: 'a',
      path: '/src/__mocks__/multi/modules/a',
    }, {
      name: 'b',
      path: '/src/__mocks__/multi/modules/b',
    }],
  });
});

Deno.test(function shouldThrowWhenBoth() {
  const dir = join(Deno.cwd(), '/src/__mocks__/invalid-multi-mixed');

  assertRejects(async () => {
    await list(dir);
  });
});

Deno.test(function shouldThrowWhenMultiModuleHasTopLevelFiles() {
  const dir = join(Deno.cwd(), '/src/__mocks__/invalid_multi_file');

  assertRejects(async () => {
    await list(dir);
  });
});

Deno.test(async function shouldUpdateAllFilesWithNewModuleVersion(t) {
  const actual = await updateVersion(
    join(Deno.cwd(), 'src/__mocks__/multi'),
    'b',
    '1.1.1',
    { dryRun: true },
  );

  actual.sort();

  await assertSnapshot(t, actual.find((file) => file.file.includes('imports')));
  await assertSnapshot(
    t,
    actual.find((file) => !file.file.includes('imports')),
  );
});
