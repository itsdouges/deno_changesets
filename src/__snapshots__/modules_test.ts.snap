export const snapshot = {};

snapshot[`shouldUpdateAllFilesWithNewModuleVersion 1`] = `
[
  {
    file: "import 'https://deno.land/x/b@1.1.1/mod.ts';
import 'https://deno.land/x/b@1.1.1/mod.ts';
import 'https://deno.land/x/b@1.1.1/main.ts';
import 'https://deno.land/x/b@1.1.1/main.ts';
import 'https://deno.land/x/b@1.1.1/utils/mod.ts';
",
    path: "/src/__mocks__/multi/modules/a/mod.ts",
  },
]
`;
